app/src/main/java/http_server/ResponseBuilder.java

```diff
@@ -1,58 +1,101 @@
 
-  public ResponseBuilder(ResponseHandler requestHandler) {
+  public ResponseBuilder(ResponseHandler requestHandler, RouteValidator routeValidator) {
     this.handler = requestHandler;
+    this.routeValidator = routeValidator;
   }
 
   public Response buildResponse(RequestReader request) {
     String path = request.getPath();
     String method = request.getMethod();
 
+    if (!routeValidator.hasPath(path)) {
+      return handler.buildUnexpectedResponse("404");
+    }
+
+    if (!routeValidator.hasMethod(path, method)) {
+      return handler.buildMethodOptionsResponse("allow", "GET, HEAD, OPTIONS", "405");
+    }
+
     if (path != null && method != null) {
       switch (path) {
-        case "/simple_get_with_body":
-          return handler.buildSimpleGetWithBodyResponse();
-
-        case "/simple_get":
-          if (method.equals("HEAD")) {
-            return handler.buildHeadResponse();
-          }
-          return handler.buildSimpleGetResponse();
-
-        case "/echo_body":
-          if (method.equals("POST")) {
-            String body = request.getBody();
-            if (body != null) {
-              return handler.buildEchoBodyResponse(body);
-            }
-          }
-          break;
-
-        case "/head_request":
-          if (method.equals("HEAD")) {
-            return handler.buildHeadResponse();
-          }
-          return handler.buildNonHeadResponse();
-
-        case "/method_options":
-          if (method.equals("OPTIONS")) {
-            return handler.buildMethodOptionsResponse("allow", "GET, HEAD, OPTIONS");
-          }
-
-        case "/method_options2":
-          if (method.equals("OPTIONS")) {
-            return handler.buildMethodOptionsResponse("allow", "GET, HEAD, OPTIONS, PUT, POST");
-          }
-
-        default:
-          break;
+        case "/simple_get_with_body" -> {
+          return buildSimpleGetWithBodyResponse(method);
+        }
+        case "/simple_get" -> {
+          return buildSimpleGetResponse(method);
+        }
+        case "/echo_body" -> {
+          return buildEchoBodyResponse(method, request.getBody());
+        }
+        case "/head_request" -> {
+          return buildHeadRequestResponse(method);
+        }
+        case "/method_options" -> {
+          return buildMethodOptionsResponse(method, "GET, HEAD, OPTIONS");
+        }
+        case "/method_options2" -> {
+          return buildMethodOptionsResponse(method, "GET, HEAD, OPTIONS, PUT, POST");
+        }
+        case "/redirect" -> {
+          return buildRedirectResponse(method);
+        }
+        default -> {
+        }
       }
     }
 
-    return handler.buildUnexpectedResponse();
+    return handler.buildUnexpectedResponse("405");
+  }
+
+  private Response buildSimpleGetWithBodyResponse(String method) {
+    if (method.equals("HEAD")) {
+      return handler.buildHeadResponse("200");
+    } else {
+      return handler.buildResponse("200", "Hello world");
```

Comment:
We spoke about this during our pairing session but just writing it down here for others to read. 

There are now two `buildResponse` methods -- one coming from the "ResponseHandler" and one in this file "ResponseBuilder".   We'll need to think about what each class's actual responsibility is and may need to update the names/methods to reflect that.  At the end of our pairing session, we were thinking that the class names should actually be switched.

This file seems to be more of the "handler" or "controller" and the other file seems to more of the builder.

It seemed like a bigger haul so Rian and I talked about deferring that potential refactor for a later time.

Reply:
sounds good, nice job identifying that